src/Network.ts

```diff
@@ -1,104 +1,109 @@
+  peers: Map<PeerId, NetworkPeer>
+  peerQ: Queue<NetworkPeer>
   swarm?: Swarm
+  joinOptions?: JoinOptions
 
-  constructor(selfId: PeerId, store: FeedStore) {
+  constructor(selfId: PeerId) {
     this.selfId = selfId
-    this.store = store
-    this.joined = new Map()
+    this.joined = new Set()
+    this.pending = new Set()
     this.peers = new Map()
+    this.peerQ = new Queue('Network:peerQ')
+    this.joinOptions = { announce: true, lookup: true }
   }
 
-  join(feedId: FeedId): void {
-    const id = discoveryId(feedId)
-    if (this.joined.has(id)) return
-    if (this.swarm) this.swarm.join(decodeId(id))
+  join(discoveryId: DiscoveryId): void {
+    if (this.swarm) {
+      if (this.joined.has(discoveryId)) return
 
-    this.joined.set(id, feedId)
+      this.joined.add(discoveryId)
+      this.swarm.join(decodeId(discoveryId), this.joinOptions)
+      this.pending.delete(discoveryId)
+    } else {
+      this.pending.add(discoveryId)
+    }
   }
 
-  leave(feedId: FeedId): void {
-    const id = discoveryId(feedId)
-    if (!this.joined.has(id)) return
-    if (this.swarm) this.swarm.leave(decodeId(id))
-    this.joined.delete(id)
+  leave(discoveryId: DiscoveryId): void {
+    this.pending.delete(discoveryId)
+    if (!this.joined.has(discoveryId)) return
+
+    if (this.swarm) this.swarm.leave(decodeId(discoveryId))
+    this.joined.delete(discoveryId)
   }
 
-  setSwarm(swarm: Swarm): void {
+  setSwarm(swarm: Swarm, joinOptions?: JoinOptions): void {
     if (this.swarm) throw new Error('Swarm already exists!')
 
+    if (joinOptions) this.joinOptions = joinOptions
     this.swarm = swarm
+    this.swarm.on('connection', this.onConnection)
+
+    for (const discoveryId of this.pending) {
+      this.join(discoveryId)
+    }
+  }
 
-    for (let id of this.joined.keys()) {
-      this.swarm.join(decodeId(id))
+  get closedConnectionCount(): number {
+    let count = 0
+    for (const peer of this.peers.values()) {
+      count += peer.closedConnectionCount
     }
+    return count
   }
 
-  close(): Promise<void> {
+  async close(): Promise<void> {
+    this.peers.forEach((peer) => {
+      peer.close()
+    })
+
     return new Promise((res) => {
-      if (!this.swarm) return res()
-      this.swarm.discovery.removeAllListeners()
-      this.swarm.discovery.close()
-      this.swarm.peers.forEach((p: any) => p.connections.forEach((con: any) => con.destroy()))
-      this.swarm.removeAllListeners()
-      res()
+      this.swarm ? this.swarm.destroy(res) : res()
     })
   }
 
-  onConnect = (peerInfo: any) => {
-    const protocol = HypercoreProtocol({
-      live: true,
-      id: decodePeerId(this.selfId),
-      encrypt: false,
-      timeout: 10000,
-      extensions: DocumentBroadcast.SUPPORTED_EXTENSIONS,
+  getOrCreatePeer(peerId: PeerId) {
+    return getOrCreate(this.peers, peerId, () => {
+      const peer = new NetworkPeer(this.selfId, peerId)
+
+      peer.connectionQ.subscribe((_conn) => {
```

Comment:
Trying to think of where the right place to document this behavior is, "We only present a peer after an initial connection has been established." Inline comments here don't really meet the need, which is "what are the characteristics of this thing?".